refactored CLI binary location to be more resiliant

benjaminwestern opened the pull request:
TL;DR: Path Detection Overhaul

This PR fixes executable path detection for tools installed via managers (`asdf`, `mise`, etc.) by replacing the simple `which` command with a more robust system.

**Please review the following key areas:**

1.  **Detection Logic (`src/utils.ts`):**
    A new 3-tier detection priority has been implemented:
    1.  User-configured path (new setting)
    2.  System `PATH` search
    3.  Common tool manager locations (Homebrew, asdf, etc.)

Is this search order robust? Are we missing any common installation paths?

2.  **Caching Strategy (`src/utils.ts`):**
    An in-memory, 5-minute cache was added to prevent repeated filesystem lookups.
Is a 5-minute cache duration a reasonable default?

3.  **Configuration (`package.json`):**
    Added optional `dataformExecutablePath` and `gcloudExecutablePath` settings for manual overrides.
    
Are the setting names and descriptions clear for end-users?

4.  **Error Handling (`src/extension.ts`):**
    "Not found" errors are now suppressed during normal operation and only shown on activation if the executable is required.
    
Does this improve the user experience without hiding critical issues?

**There are no breaking changes.** This is a backward-compatible enhancement.

ashish10alex commented on src/utils.ts:
```diff
@@ -463,20 +463,213 @@ export async function fetchGitHubFileContent(): Promise<string> {
+            windowsHide: true
+        }).trim();
+
+        if (result) {
```
could we please add `logger.debug` statements here and other relevant places which might help us with debugging user issues later on. E.g.  when cache is used we see a log when debug mode is enabled

ashish10alex commented on src/utils.ts:
```diff
@@ -463,20 +463,213 @@ export async function fetchGitHubFileContent(): Promise<string> {
+// Find executable using built-in detection + user overrides
+function findExecutableInPaths(executableName: string): string | null {
+    const cacheKey = `${executableName}:${process.platform}`;
+    const cached = executablePathCache.get(cacheKey);
```
I was unable to verify if the cache was working. I ran the extension in debug mode, it found the path as expected. Then did `read window` in the same debug session, however the cached variable was undefined. Can you please clarify if this is expected ?

ashish10alex reviewed (changes requested):
Thanks @benjaminwestern for yet another contribution :)

benjaminwestern commented:
@ashish10alex 
**Binary Path Resolution:**
- Added debug logs for cache hits/misses, user config paths, system PATH searches, and common location searches for dataform/gcloud executables

**Compilation Cache:**
- Added logs for cache usage vs fresh compilation, cache population/clearing, and dataform CLI path resolution

**Extension Lifecycle:**
- Added logs for extension activation, cache initialization, and executable availability checks

**Cache Behavior:**
- Cache being undefined after "reload window" is expected - VS Code extensions lose memory state on reload
